torch/csrc/jit/mobile/versioned_operators.cpp

```diff
@@ -0,0 +1,185 @@
+      "The version number of ",
+      c10::toString(name),
+      ", ",
+      op_version,
+      ", is smaller than the minimum version number, ",
+      container.min_version(),
+      ", supported in this runtime. The model file is too old. ",
+      "Update the model with most recent code.");
+
+  // Forward compatibility breakage
+  TORCH_CHECK(
+      op_version <= container.max_version(),
+      "The version number of ",
+      c10::toString(name),
+      ", ",
+      op_version,
+      ", is larger than the maximum version number, ",
+      container.max_version(),
+      ", supported in this runtime. Update the runtime to be compatible",
+      " to this model.");
+
+  for (const auto& info : container.version_vec) {
+    if (info.version == op_version) {
+      return info;
+    }
+  }
+
+  if (op_version == container.max_version() + 1) {
+    // Current version
+    return VersionInfo(op_version, name, c10::nullopt);
+  }
+
+  // The situation where op_version is in [min_version, max_version + 1], but
+  // not in the container, which should not happen.
+  TORCH_CHECK(
+      true,
+      "The version number of ",
+      c10::toString(name),
+      ", ",
+      op_version,
+      " is not compatible in this runtime. ");
+
+  return VersionInfo(
+      0, c10::OperatorName("aten::_convolution", ""), c10::nullopt);
+}
+
+static const std::unordered_map<std::string, VersionInfoContainer>
+    op_version_table(
+        {{"aten::_convolution",
+          VersionInfoContainer(
+              {{0, c10::OperatorName("aten::_convolution", ""), true}})}});
+} // namespace
+
+OperatorFunctor findOperatorFromName(const c10::OperatorName& opname) {
+  auto jit_op = findOperatorFor(opname);
+  OperatorFunctor fn;
+  if (jit_op) {
+    fn = [jit_op](Stack& stack) { jit_op->getOperation()(&stack); };
+  } else {
+    auto op = c10::Dispatcher::singleton().findSchema(opname);
+    if (op.has_value()) {
+      fn = [op](Stack& stack) { op->callBoxed(&stack); };
+    } else {
+      return fn;
+    }
+  }
+  return nullptr;
+}
+
+OperatorFunctor operator_resolver(
+    const c10::OperatorName& opname,
+    int64_t op_version,
+    int64_t model_version) {
+  if (model_version > 0x3LL) {
+    auto it = op_version_table.find(toString(opname));
+    if (it == op_version_table.end()) {
+      // Not in the version table, by default it fall through the original
+      // opname with no compatibility treatment
+      return findOperatorFromName(opname);
+    } else {
+      auto opinfo = getVersioninfo(it->second, opname, op_version);
+      auto fn = findOperatorFromName(opinfo.name);
+      if (opinfo.default_val) {
+        fn = [fn, opinfo](Stack& stack) {
+          stack.push_back(opinfo.default_val.value());
+          fn(stack);
+        };
+      }
+      return fn;
+    }
+  } else if (model_version == 0x3LL) {
+    auto fn = findOperatorFromName(opname);
+    if (opname == c10::OperatorName("aten::_convolution", "")) {
```

Comment:
I'm confused why this special case for convolution is needed here. Shouldn't its entry in op_version_table take care of its handling?

Reply:
Models with version 3 does not have the operator version table. Since we can't get op version info from the model, we have to do some manual check.

Reply:
Would you elaborate on this?

If this code is reading an older model then doesn't it still have access to the operator version table because the table is part of the reader's code base?

Reply:
I think what Martin means is that for this versioning mechanism to work there needs to be a version associated to the operators in the model as well (to be implemented in a subsequent PR).

Since that's not going to be present in old models, then he needs to fallback. to the old "backward compatibility logic".